File: grpc-java-examples/src/main/java/com/example/grpc/headers/HeadersServerInterceptor.java
package com.example.grpc.headers;

import io.grpc.ForwardingServerCall.SimpleForwardingServerCall;
import io.grpc.Metadata;
import io.grpc.ServerCall;
import io.grpc.ServerCallHandler;
import io.grpc.ServerInterceptor;

import java.util.logging.Level;
import java.util.logging.Logger;

class HeadersServerInterceptor implements ServerInterceptor {

    private static final Logger logger = Logger.getLogger(HeadersServerInterceptor.class.getName());
    private static final Metadata.Key<String> CUSTOM_HEADER_KEY =
        Metadata.Key.of("custom_server_header", Metadata.ASCII_STRING_MARSHALLER);

    @Override
    public <ReqT, RespT> ServerCall.Listener<ReqT> interceptCall(
        ServerCall<ReqT, RespT> call,
        Metadata requestHeaders,
        ServerCallHandler<ReqT, RespT> next) {
        logger.log(Level.INFO, "header received from client: {0}", requestHeaders);
        return next.startCall(new SimpleForwardingServerCall<>(call) {
            @Override
            public void sendHeaders(Metadata responseHeaders) {
                responseHeaders.put(CUSTOM_HEADER_KEY, "custom_response_value");
                super.sendHeaders(responseHeaders);
            }
        }, requestHeaders);
    }
}

File: grpc-kotlin-examples/src/main/kotlin/com/example/grpc/headers/HeadersServerInterceptor.kt
package com.example.grpc.headers

import io.grpc.ForwardingServerCall.SimpleForwardingServerCall
import io.grpc.Metadata
import io.grpc.ServerCall
import io.grpc.ServerCallHandler
import io.grpc.ServerInterceptor
import java.util.logging.Level
import java.util.logging.Logger

internal class HeadersServerInterceptor : ServerInterceptor {
  override fun <ReqT, RespT> interceptCall(
    call: ServerCall<ReqT?, RespT?>?,
    requestHeaders: Metadata?,
    next: ServerCallHandler<ReqT?, RespT?>
  ): ServerCall.Listener<ReqT?>? {
    logger.log(Level.INFO, "header received from client: {0}", requestHeaders)
    return next.startCall(object : SimpleForwardingServerCall<ReqT?, RespT?>(call) {
      override fun sendHeaders(responseHeaders: Metadata) {
        responseHeaders.put<String?>(CUSTOM_HEADER_KEY, "custom_response_value")
        super.sendHeaders(responseHeaders)
      }
    }, requestHeaders)
  }

  companion object {
    private val logger: Logger = Logger.getLogger(HeadersServerInterceptor::class.java.getName())
    private val CUSTOM_HEADER_KEY: Metadata.Key<String?> =
      Metadata.Key.of<String?>("custom_server_header", Metadata.ASCII_STRING_MARSHALLER)
  }
}
